Added YOLO for detecting bounding boxes per frame on an input video

hargunkaur286 opened the pull request:
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/neuroinformatics-unit/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

This PR adds the ability to run a trained object detector (YOLO) on an input video file and outputs per-frame bounding box detections into a JSON file. This functionality was outlined as part of the PR(#12)

**What does this PR do?**

-  added the inference logic for YOLO into a dedicated file (yolo_inference.py) under the inference folder.
- run_detector.py acts as a CLI entrypoint. Users can specify:
        --video_path:  (I’ve attached a sample video for now)
        --model_path: path to the YOLO weights file (I used yolov8n.pt)
        --output_path: path to save the generated detections.json
        --visualize: optional flag to enable real-time bounding box display

I wanted to get your feedback on the following:
1. Does this PR align with what was expected for YOLO? If so, I’ll proceed with FasterRCNN.
2. Should we consider standardizing outputs (e.g., bounding box format, class names)?
3. Should we move the sample video to a more appropriate samples/ or assets/ directory in the long run?

For FasterRCNN:
I already have a working version in progress, but I want to confirm whether we want it loaded from a .pth file or if it's okay to use torchvision’s built-in pretrained model. Once that’s clear, I’ll follow the same format and add the tests.


https://github.com/user-attachments/assets/fd897561-3f87-4a3b-a71f-e2f6acfa48e1

## References

Fixes #12 

## How has this PR been tested?
Tested locally using this command in the ethology folder

> python -m detectors.inference.run_detector \
  --video_path video/ethology_sample.mp4 \
  --detector_type yolo \
  --model_path yolov8n.pt \
  --output_path detections.json \
  --visualize
 
## Is this a breaking change?

No breaking changes as this is a new feature and the file is isolated under the inference folder.

## Does this PR require an update to the documentation?

If any features have changed, or have been added. Please explain how the
documentation has been updated.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

sfmig commented:
Hi @hargunkaur286, thanks for having a go!

Would you mind having a go at fixing the CI failures first? You may find our [contributing](https://ethology.neuroinformatics.dev/community/contributing.html#how-to-contribute) guide useful for this.

Also remember that all added features should be tested before marking as ready for review (as per the checklist in the description).

hargunkaur286 commented:
> Hi @hargunkaur286, thanks for having a go!
> 
> Would you mind having a go at fixing the CI failures first? You may find our [contributing](https://ethology.neuroinformatics.dev/community/contributing.html#how-to-contribute) guide useful for this.
> 
> Also remember that all added features should be tested before marking as ready for review (as per the checklist in the description).

Hey @sfmig can you have a look now?

sfmig commented:
Hi @hargunkaur286, the CI checks for the tests and the doc building appear to still be failing. 

For the tests, you can see from the logs [here](https://github.com/neuroinformatics-unit/ethology/actions/runs/14260133856/job/39989026995?pr=78) that the `ultralytics` dependency is missing. Make sure that any added dependencies are added correctly to pyproject.toml.

Would you mind verifying if this fixes the tests CI check, and if not investigating further? You can follow a similar strategy to diagnose why the docs CI check is failing.

hargunkaur286 commented:
> Hi @hargunkaur286, the CI checks for the tests and the doc building appear to still be failing.
> 
> For the tests, you can see from the logs [here](https://github.com/neuroinformatics-unit/ethology/actions/runs/14260133856/job/39989026995?pr=78) that the `ultralytics` dependency is missing. Make sure that any added dependencies are added correctly to pyproject.toml.
> 
> Would you mind verifying if this fixes the tests CI check, and if not investigating further? You can follow a similar strategy to diagnose why the docs CI check is failing.

Hey @sfmig can you please check now. I believe all the checks are passing now, including the [tests](https://github.com/hargunkaur286/ethology/actions/runs/14389673330) and the [docs](https://github.com/hargunkaur286/ethology/actions/workflows/docs_build_and_deploy.yml) ones.

hargunkaur286 commented:
> Hi @hargunkaur286,
> 
> Thanks for looking into this further. The [latest checks](https://github.com/neuroinformatics-unit/ethology/pull/78/checks) still show errors, for the docs building and for code coverage.
> 
> I would try to fix the docs building first. I noticed you added `"ultralytics"` as a dependency in pyproject.toml, but it is added as an optional dependency, rather than as a dependency. It should be added under the section called [dependencies](https://github.com/hargunkaur286/ethology/blob/414cfdd8e74f5287cfeee9ff5399b93adb7b220b/pyproject.toml#L21). That may fix some of the issues.

Hello @sfmig, yes you're right, I had added it under the optional dependencies. But now, I've moved it under dependencies. Sorry for the constant trouble :') but is there anything else I can do to improve?

sfmig commented:
Hi @hargunkaur286, thanks for your work!

I noticed some of the latest CI checks are still failing - specifically the code coverage one, which indicates that the added code is not sufficiently tested. 

I acknowledge it is especially challenging to work in a package like `ethology` in which we are still figuring out how to best implement things and don't have much codebase to use as reference. But it is my impression that this work comes a bit too early in the life of this package - we will be probably better placed to address this when we have a more established codebase, especially re the annotations and the dataset modules. 

At the moment, taking this PR to completion will require a lot of work and probably several PRs and discussions. For example, we are not yet set on whether we want ethology to have a CLI interface and we would need to discuss if we want to depend on `ultralytics`, since it has a GPL license. These are just some of the tricky bits, and for these reasons I would suggest closing this PR for now.

As you know, contributions to our NIU tools are always welcome, and you may find some of our other issues more attainable, especially in our more mature tools like brainglobe or movement (see the full list of neuroinformatics team repositories [here](https://github.com/neuroinformatics-unit)).
